### PR TITLE
[Driver][SYCL] Issue a warning for converting 'c' input to 'c++' in SYCL mode

### PR DESCRIPTION
**Pull request by hchilama**

### CODE REVIEW

File: clang/include/clang/Basic/DiagnosticGroups.td [hunk 1]
```diff
@@ -1239,6 +1239,7 @@ def Sycl2020Compat : DiagGroup<"sycl-2020-compat">;
 def SyclStrict : DiagGroup<"sycl-strict", [ Sycl2017Compat, Sycl2020Compat]>;
 def SyclTarget : DiagGroup<"sycl-target">;
 def SyclFPGAMismatch : DiagGroup<"sycl-fpga-mismatch">;
+def SyclCompat : DiagGroup<"sycl-compat">;
```
Comment (mdtoguchi) [inline clang/include/clang/Basic/DiagnosticGroups.td:1242]: I'm on the fence on the naming of the `-W` option here.

**Review by mdtoguchi — DISMISSED** [review]
Driver looks OK - but would like to see opinions on the diagnostic option name from the CFE folks.

File: clang/include/clang/Basic/DiagnosticGroups.td (hunk 1 above)
Comment (erichkeane) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: Not up to me, but I agree here.  I don't think this warning group fits well here.  I'd suggest the CFE/driver reviewers see if we can bikeshed this better

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 2]
```diff
@@ -669,4 +669,6 @@ def err_drv_cuda_offload_only_emit_bc : Error<
 def warn_drv_jmc_requires_debuginfo : Warning<
   "/JMC requires debug info. Use '/Zi', '/Z7' or other debug options; option ignored">,
   InGroup<OptionIgnored>;
+def warn_drv_fsycl_with_c_type : Warning<
+  "treating '%0' input as '%1' when -fsycl is used">, InGroup<SyclCompat>;
```
Comment (elizabethandrews) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:673]: I'm not sure if we need to mention `when -fsycl is used` here. @premanandrao WDYT?

File: clang/include/clang/Basic/DiagnosticGroups.td (hunk 1 above)
Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: I'm not sure what the group should be named here. `InvalidSourceFile`/`UnsupportedFileType`/`LanguageMismatch`? I don't really like any of these. @premanandrao do you have a suggestion?

Comment (hchilama) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: > I did quick search to see if I can find a better Warning group and I saw 2 other diagnostics which look related -
> 
> ```
> def err_drv_fsycl_with_c_type : Error<
>   "'%0' must not be used in conjunction with '-fsycl', which expects C++ source">;
> 
> def warn_drv_treating_input_as_cxx : Warning<
>   "treating '%0' input as '%1' when in C++ mode, this behavior is deprecated">,
>   InGroup<Deprecated>;
> ```
> 
> I assume the latter is used for non-sycl programs but would you know what the former is used for now?

Hi @elizabethandrews , In above 2 defs "err_drv_fsycl_with_c_type" can only be used for error message. so I can't use it for a warning. And the other one "warn_drv_treating_input_as_cxx" I tried using it but @mdtoguchi said "The warning is not accurate for -fsycl". So I have created new one.

Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: Sorry for not being clear.  My question was - when is `err_drv_fsycl_with_c_type ` thrown ? I was curious about warning vs error functionality here.

Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: Right, but what is the scenario in which error is thrown? If we are converting C types to C++ types for SYCL, why does that error diagnostic exist.

Comment (hchilama) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: This error emits when 'c' compilation is forced in -fsycl mode.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td (hunk 2 above)
Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticDriverKinds.td:673]: I think there is some value is saying why the input type was changed.  Suggest quotes around -fsycl though:
```suggestion
  "treating '%0' input as '%1' when '-fsycl' is used">, InGroup<SyclCompat>;
```

I see this existing driver diagnostic as something that has a similar format:
```
 warn_drv_treating_input_as_cxx : Warning<
  "treating '%0' input as '%1' when in C++ mode, this behavior is deprecated">,
  InGroup<Deprecated>;

```

File: clang/include/clang/Basic/DiagnosticGroups.td (hunk 1 above)
Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: @hchilama I'm ok with `InvalidSourceFile` or `ExpectedFileType`. If you agree, please change the group name. Thanks!

Comment (hchilama) [reply clang/include/clang/Basic/DiagnosticGroups.td:1242]: Sure thanks @elizabethandrews, I will update the group name.

File: clang/include/clang/Basic/DiagnosticGroups.td [hunk 3]
```diff
@@ -275,6 +275,9 @@ def CPre2xCompat : DiagGroup<"pre-c2x-compat">;
 def CPre2xCompatPedantic : DiagGroup<"pre-c2x-compat-pedantic",
                                      [CPre2xCompat]>;
 
+// Warning group for treating C input as C++ input
```
Comment (smanna12) [inline clang/include/clang/Basic/DiagnosticGroups.td:278]: LGTM.
Minor nit
```suggestion
// Warning group for treating C input as C++ input.
```

**Comment by smanna12** [comment]
@hchilama, you have conflicts. could you please resolve that? Thanks

**Review by premanandrao — APPROVED** [review]
I am okay with the changes.

**Review by smanna12 — DISMISSED** [review]

File: clang/include/clang/Basic/DiagnosticGroups.td [hunk 4]
```diff
@@ -275,6 +275,9 @@ def CPre2xCompat : DiagGroup<"pre-c2x-compat">;
 def CPre2xCompatPedantic : DiagGroup<"pre-c2x-compat-pedantic",
                                      [CPre2xCompat]>;
 
+// Warning for treating C input as C++ input.
+def ExpectedFileType : DiagGroup<"ExpectedFileType">;
```
Comment (mdtoguchi) [inline clang/include/clang/Basic/DiagnosticGroups.td:279]: The name of the diagnostic when used as an option probably should be something like:  `DiagGroup<"expected-file-type">`

Comment (hchilama) [reply clang/include/clang/Basic/DiagnosticGroups.td:279]: Yes, I have verified the other group naming. I will update this.

**Review by smanna12 — APPROVED** [review]

**Review by mdtoguchi — APPROVED** [review]

**Review by premanandrao — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

**Pull request merged**